src/main/java/org/opensearch/geospatial/ip2geo/action/PutDatasourceTransportAction.java

```diff
@@ -0,0 +1,189 @@
+import org.opensearch.geospatial.ip2geo.common.DatasourceManifest;
+import org.opensearch.geospatial.ip2geo.common.DatasourceState;
+import org.opensearch.geospatial.ip2geo.common.GeoIpDataHelper;
+import org.opensearch.geospatial.ip2geo.common.Ip2GeoSettings;
+import org.opensearch.geospatial.ip2geo.jobscheduler.Datasource;
+import org.opensearch.geospatial.ip2geo.jobscheduler.DatasourceExtension;
+import org.opensearch.index.engine.VersionConflictEngineException;
+import org.opensearch.tasks.Task;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.TransportService;
+
+/**
+ * Transport action to create datasource
+ */
+@Log4j2
+public class PutDatasourceTransportAction extends HandledTransportAction<PutDatasourceRequest, AcknowledgedResponse> {
+    private final Client client;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+
+    private TimeValue timeout;
+    private int indexingBulkSize;
+
+    /**
+     * Default constructor
+     * @param transportService the transport service
+     * @param actionFilters the action filters
+     * @param client the client
+     * @param clusterService the cluster service
+     * @param threadPool the thread pool
+     * @param settings the settings
+     * @param clusterSettings the cluster settings
+     */
+    @Inject
+    public PutDatasourceTransportAction(
+        final TransportService transportService,
+        final ActionFilters actionFilters,
+        final Client client,
+        final ClusterService clusterService,
+        final ThreadPool threadPool,
+        final Settings settings,
+        final ClusterSettings clusterSettings
+    ) {
+        super(PutDatasourceAction.NAME, transportService, actionFilters, PutDatasourceRequest::new);
+        this.client = client;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        timeout = Ip2GeoSettings.TIMEOUT_IN_SECONDS.get(settings);
+        clusterSettings.addSettingsUpdateConsumer(Ip2GeoSettings.TIMEOUT_IN_SECONDS, newValue -> timeout = newValue);
+        indexingBulkSize = Ip2GeoSettings.INDEXING_BULK_SIZE.get(settings);
+        clusterSettings.addSettingsUpdateConsumer(Ip2GeoSettings.INDEXING_BULK_SIZE, newValue -> indexingBulkSize = newValue);
+    }
+
+    @Override
+    protected void doExecute(final Task task, final PutDatasourceRequest request, final ActionListener<AcknowledgedResponse> listener) {
+        try {
+            Datasource jobParameter = Datasource.Builder.build(request);
+            IndexRequest indexRequest = new IndexRequest().index(DatasourceExtension.JOB_INDEX_NAME)
+                .id(jobParameter.getId())
+                .source(jobParameter.toXContent(JsonXContent.contentBuilder(), null))
+                .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+                .opType(DocWriteRequest.OpType.CREATE);
+            client.index(indexRequest, new ActionListener<>() {
```

Comment:
See https://github.com/opensearch-project/k-NN/pull/859/files for interacting with system index when security is enabled. Maybe take this for a future PR

Reply:
Thanks